src/rgw/rgw_notify.cc

```diff
@@ -179,25 +181,88 @@ class Manager {
+    if (now < next_cleanup) {
+      return 0;
+    }
+    const auto stale_timestamp = now - std::chrono::seconds(stale_reservations_period_s);
+    next_cleanup = now + std::chrono::seconds(reservations_cleanup_period_s);
+    const auto tp = ceph::coarse_real_time::clock::to_time_t(next_cleanup);
+    ldout(cct, 20) << "INFO: stale reservation cleanup for queue: " << queue_name 
+      << ". next cleanup will happen at: " << std::ctime(&tp) << dendl;
+    librados::ObjectReadOperation list_op;
+    bufferlist obl;
+    int rval;
+    cls_2pc_queue_list_reservations(list_op, &obl, &rval);
+    auto ret = rgw_rados_operate(rados_ioctx, queue_name, &list_op, nullptr, optional_yield(io_context, yield));
+    if (ret < 0) {
+      ldout(cct, 5) << "WARNING: failed to cleanup stale reservation from queue: " << queue_name
+        << ". failed to get list, error: " << ret << dendl;
+      return 0;
+    }
+
+    cls_2pc_reservations reservations;
+    ret = cls_2pc_queue_list_reservations_result(obl, reservations);
+    if (ret < 0) {
+      ldout(cct, 5) << "WARNING: failed to cleanup stale reservation from queue: " << queue_name
+      << ". failed to parse list, error: " << ret << dendl;
+      return 0;
+    }
+    
+    auto cleaned_reservations = 0U;
+    for (const auto& r : reservations) {
+      if (r.second.timestamp < stale_timestamp) {
+        librados::ObjectWriteOperation abort_op;
+        cls_2pc_queue_abort(abort_op, r.first);
+        ret = rgw_rados_operate(rados_ioctx, queue_name, &abort_op, optional_yield(io_context, yield));
+        if (ret < 0) {
+          ldout(cct, 5) << "WARNING: failed to cleanup stale reservation from queue: " << queue_name
+            << ". failed to abort: " << r.first <<  ", error: " << ret << dendl;
+        } else {
+          ++cleaned_reservations;
+          const auto tp = ceph::coarse_real_time::clock::to_time_t(r.second.timestamp);
+          ldout(cct, 5) << "WARNING: stale reservetion: " << r.first << " with timestamp: " << std::ctime(&tp) 
+            <<  " cleaned from queue: " << queue_name << dendl;
+        }
+      }
+    }
+    return cleaned_reservations;
+  }
+
   // processing of a specific queue
   void process_queue(const std::string& queue_name, const bool& owned, spawn::yield_context yield) {
     constexpr auto max_elements = 1024;
     auto is_idle = false;
     const std::string start_marker;
+    auto next_cleanup = ceph::coarse_real_time::clock::now();
     while (owned) {
+
+      // if queue was empty the last time, sleep for idle timeout
       if (is_idle) {
         Timer timer(io_context);
         timer.expires_from_now(std::chrono::microseconds(queue_idle_sleep_us));
         boost::system::error_code ec;
 	      timer.async_wait(yield[ec]);
       }
-      is_idle = true;
 
+      // this should happen regardless of the queue being idle or not
+      const auto cleaned_reservations = clean_stale_reservation(next_cleanup, yield, queue_name);
```

Comment:
what do you think about moving this into a separate coroutine that runs concurrently? that way it can run on its own timer, and doesn't slow down queue processing

Reply:
good suggestion (I also don't like the call to "now()" in each iteration)